### PR TITLE
Support union type in casters

### LINKED ISSUE #209: Union type support
On caster class instantiation there's no check if object property has union type:
https://github.com/spatie/data-transfer-object/blob/d101e5185844eb92bf5fb8d0d1a200f1778e9802/src/Reflection/DataTransferObjectProperty.php#L85-L88
In that case it fails because ReflectionUnionType doesn't have getName() method.

### PR DESCRIPTION
**Pull request by daniser**
Closes #209

### CODE REVIEW

File: src/Reflection/DataTransferObjectProperty.php [hunk 1]
```diff
@@ -82,27 +82,17 @@ private function resolveCaster(): ?Caster
         /** @var \Spatie\DataTransferObject\Attributes\CastWith $attribute */
         $attribute = $attributes[0]->newInstance();
 
+        $typeNames = array_map(fn ($type) => $type->getName(), $this->extractTypes());
+
         return new $attribute->casterClass(
-            $this->reflectionProperty->getType()?->getName(),
+            implode('|', $typeNames),
```
Comment (brendt) [inline src/Reflection/DataTransferObjectProperty.php:88]: How would this work without making changes to the `CastWith::__construct` attribute? It still expects a single class name. I understand the issue but I don't see how this PR fixes it.

**Review by brendt — CHANGES_REQUESTED** [review]

**Comment by daniser** [comment]
CastWith attribute doesn't need any changes AFAIK, but two supplied casters do. I've replaced "|"-delimited string to type name array and made needed changes to the casters.

File: src/Reflection/DataTransferObjectProperty.php (hunk 1 above)
Comment (daniser) [reply src/Reflection/DataTransferObjectProperty.php]: `CastWith` attribute doesn't need any changes AFAIK, but two supplied casters do. I've replaced "|"-delimited string with type name array and made needed changes to the casters.

**Comment by brendt** [comment]
Can you submit tests that verify everything works as intended?

**Comment by daniser** [comment]
> Can you submit tests that verify everything works as intended?

Done!

**Comment by brendt** [comment]
Sorry for the late response.

Can you explain how this example works?

```php
    #[CastWith(RoundingCaster::class)]
    public float | int $bar;
```

`RoundingCaster` will always cast to `float`, so why would `int` still be needed there?

**Comment by daniser** [comment]
> Sorry for the late response.
> 
> Can you explain how this example works?
> 
> ```
>     #[CastWith(RoundingCaster::class)]
>     public float | int $bar;
> ```
> 
> `RoundingCaster` will always cast to `float`, so why would `int` still be needed there?

It was my oversight; fixed.

**Comment by brendt** [comment]
Thanks!

**Comment by brendt** [comment]
Tagged: https://github.com/spatie/data-transfer-object/releases/tag/3.5.0

**Comment by sidigi** [comment]
finaly

**Comment by suyar** [comment]
Good job.

**Pull request merged**